src/main/java/com/icthh/xm/uaa/security/oauth2/AuthOtpTokenGranter.java

```diff
@@ -0,0 +1,80 @@
+package com.icthh.xm.uaa.security.oauth2;
+
+import com.icthh.xm.uaa.domain.GrantType;
+import com.icthh.xm.uaa.security.DomainUserDetails;
+import com.icthh.xm.uaa.security.DomainUserDetailsService;
+import com.icthh.xm.uaa.security.oauth2.idp.source.model.XmAuthenticationToken;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.mapping.GrantedAuthoritiesMapper;
+import org.springframework.security.core.authority.mapping.NullAuthoritiesMapper;
+import org.springframework.security.oauth2.common.exceptions.InvalidGrantException;
+import org.springframework.security.oauth2.provider.ClientDetails;
+import org.springframework.security.oauth2.provider.ClientDetailsService;
+import org.springframework.security.oauth2.provider.OAuth2Authentication;
+import org.springframework.security.oauth2.provider.OAuth2RequestFactory;
+import org.springframework.security.oauth2.provider.TokenRequest;
+import org.springframework.security.oauth2.provider.token.AbstractTokenGranter;
+import org.springframework.security.oauth2.provider.token.AuthorizationServerTokenServices;
+import org.springframework.stereotype.Service;
+
+import java.util.Collection;
+import java.util.HashMap;
+
+@Slf4j
+@Service
+public class AuthOtpTokenGranter extends AbstractTokenGranter {
+
+    private static String TOKEN_REQUEST_OTP_PARAM = "auth_otp";
```

Comment:
in request I have to pass otp or auth_otp?

how about next request?

grant_type=otp&username=380932222222&otp=4812

@amedvedchuk  pls say you opinion